### ISSUE TITLE
SameDiff doesn't run on Java 12

### ISSUE THREAD

**Issue by treo** [deeplearning4j/deeplearning4j#7572]
The following simple application runs fine on Java 8 but throws an exception with Java 12.

```java
import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.linalg.factory.Nd4j;

public class App
{
    public static void main( String[] args )
    {
        SameDiff sd = SameDiff.create();
        SDVariable one = sd.var("one", Nd4j.ones(1,2,3));
        one.plus(one).sum("out");
        sd.exec(null, "out");
    }
}
```

Exception:
```
Exception in thread "main" java.lang.NullPointerException
	at org.nd4j.autodiff.functions.DifferentialFunction.propertiesForFunction(DifferentialFunction.java:157)
	at org.nd4j.autodiff.functions.DifferentialFunction.resolvePropertiesFromSameDiffBeforeExecution(DifferentialFunction.java:544)
	at org.nd4j.autodiff.samediff.internal.InferenceSession.getAndParameterizeOp(InferenceSession.java:622)
	at org.nd4j.autodiff.samediff.internal.InferenceSession.getAndParameterizeOp(InferenceSession.java:37)
	at org.nd4j.autodiff.samediff.internal.AbstractSession.output(AbstractSession.java:249)
	at org.nd4j.autodiff.samediff.SameDiff.exec(SameDiff.java:3932)
	at tech.dubs.App.main(App.java:27)
```

The issue is due to Guavas ClassPath returning an empty List in https://github.com/deeplearning4j/deeplearning4j/blob/master/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/imports/converters/DifferentialFunctionClassHolder.java#L167 on the newer java version.

According to https://github.com/google/guava/issues/3249 we would have to change to use both `ClassLoader.getSystemClassLoader()` here AND use a newer guava version (min 23.1).

As Guava > 20 requires Java 8, and has been split into JRE and Android artifacts, and we don't have full classpath control in environments like Spark, it may be a good idea to investigate a different solution.

**Comment by AlexDBlack** [comment]
Related: https://github.com/deeplearning4j/deeplearning4j/issues/7170

OK, so what we need here is basically a mapping between TF op names and Classes.
That's currently done at runtime using reflection.

We can't use Reflection library (as we used to) because of Android.
As you've pointed out, we can't rely on just upgrading Guava, because if an earlier version leaks onto user classpath, SameDiff will break.

Some options here:
1. We scan at build time, and load the results from an xml/json file (no runtime scanning): https://github.com/ronmamo/reflections#integrating-into-your-build-lifecycle
2. We implement some sort of manual mapping (say, a file that just populates a Map<String,Class>). The only issue here is coverage, forgetting to add stuff to this manual class. However, there's nothing stopping us building a simple coverage test using (test scope) Reflections library.

**Comment by saudet** [comment]
BTW, when is this needed? IIRC, Alex solved a similar issue with Jackson and JSON serialization by using the fully qualified name of classes to associate information with them. This way whenever we load or save something, we don't need to scan anything, we just need to query the class loader with the FQN.

**Comment by AlexDBlack** [comment]
@saudet I'd like to try to get this into the release if possible. I don't think it'll be especially hard to solve, but we'll see.
Slightly different problem here vs. Jackson. Unlike with JSON, don't control the source format (i.e., when reading tensorflow protobuf), so we still need that TF op name to SameDiff op class mapping defined somewhere.

**Comment by treo** [comment]
This also isn't just about tensorflow though, it loads all of the other custom ops defined in libnd4j that way as well.

**Comment by AlexDBlack** [comment]
libnd4j op names doesn't rely on reflections - it's a call from c++ via `Nd4j.getExecutioner().getCustomOperations()`

https://github.com/deeplearning4j/deeplearning4j/blob/bef25f97bbfa4e4cf3352af65cfba308bb7cf852/nd4j/nd4j-backends/nd4j-backend-impls/nd4j-native/src/main/java/org/nd4j/linalg/cpu/nativecpu/ops/NativeOpExecutioner.java#L1434-L1470

**Comment by treo** [comment]
When DifferentialFunctionClassHolder is instantiated it gets all available ops via reflection and populates `fieldsForFunction` by iterating over the classes it gets that way. 

Then when looking them up, it looks at the `fieldsForFunction` field.
https://github.com/deeplearning4j/deeplearning4j/blob/master/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/functions/DifferentialFunction.java#L154 

In my example it fails because it can't find the properties for `add`, which is hardly a tensorflow only op.

**Comment by AlexDBlack** [comment]
OK, so we need to enumerate all ops for two reasons, not one. That doesn't change the underlying issue (enumerate all ops without reflection) here. :)

**Comment by treo** [comment]
Sure, just wanted to ensure the focus isn't only on Tensorflow (which does get some enumeration via `TensorflowDescriptorParser.opDescs()` already.

https://github.com/deeplearning4j/deeplearning4j/blob/master/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/imports/converters/DifferentialFunctionClassHolder.java#L287-L294

**Comment by AlexDBlack** [comment]
That's different enumeration using a different mechanism.

Anyway, I'll take this one. I already know what's going on here, why, and how to fix it.

**Comment by AlexDBlack** [comment]
Fix is merged. https://github.com/deeplearning4j/deeplearning4j/pull/7575
I went with the "manually created list of classes" approach to replace guava reflections.
It's simple/robust/fast, and we have now have tests to ensure we actually keep the class list up to date.

**Comment by treo** [comment]
Great! 

I guess all of those loss functions weren't actually proper op mappings?

**Comment by AlexDBlack** [comment]
Yes, there was the intent early in SameDiff's existence to import to those existing ILossFunction instances.
But they don't work as SameDiff functions and never have.
